src/main/java/com/timgroup/statsd/StatsDNonBlockingProcessor.java

```diff
@@ -0,0 +1,121 @@
+            final int maxPacketSizeBytes, final int poolSize, final int workers)
+            throws Exception {
+
+        super(queueSize, handler, maxPacketSizeBytes, poolSize, workers);
+        this.qSize = new AtomicInteger(0);
+        this.qCapacity = queueSize;
+        this.messages = new ConcurrentLinkedQueue<String>();
+    }
+
+    @Override
+    boolean send(final String message) {
+        if (!shutdown) {
+            if (qSize.get() < qCapacity) {
+                messages.offer(message);
+                qSize.incrementAndGet();
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    @Override
+    public void run() {
+
+        for (int i=0 ; i<workers ; i++) {
+            executor.submit(new Runnable() {
+                public void run() {
+                    boolean empty;
+                    ByteBuffer sendBuffer;
+
+                    try {
+                        sendBuffer = bufferPool.borrow();
+                    } catch(final InterruptedException e) {
+                        handler.handle(e);
+                        return;
+                    }
+
+                    while (!((empty = messages.isEmpty()) && shutdown)) {
+
+                        try {
+                            if (empty) {
+                                Thread.sleep(WAIT_SLEEP_MS);
+                                continue;
+                            }
+
+                            if (Thread.interrupted()) {
+                                return;
+                            }
+                            final String message = messages.poll();
+                            if (message != null) {
+                                qSize.decrementAndGet();
+                                final byte[] data = message.getBytes(MESSAGE_CHARSET);
+                                if (sendBuffer.capacity() < data.length) {
+                                    throw new InvalidMessageException(MESSAGE_TOO_LONG, message);
+                                }
+                                if (sendBuffer.remaining() < (data.length + 1)) {
+                                    outboundQueue.put(sendBuffer);
+                                    sendBuffer = bufferPool.borrow();
+                                }
+                                if (sendBuffer.position() > 0) {
+                                    sendBuffer.put((byte) '\n');
+                                }
+                                sendBuffer.put(data);
+                                if (null == messages.peek()) {
+                                    outboundQueue.put(sendBuffer);
+                                    sendBuffer = bufferPool.borrow();
+                                }
+                            }
+                        } catch (final InterruptedException e) {
+                            if (shutdown) {
+                                return;
```

Comment:
Do we need `endSignal.countDown();` as for https://github.com/DataDog/java-dogstatsd-client/pull/96/files#diff-d87f261783303f309bda2086a2e14756R86

Reply:
nice catch, yup!